### PR TITLE
add elfio

### PR DESCRIPTION
**Pull request by a4z**
Specify library name and version:  **elfio/3.8**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by SSE4 — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (prince-chrismc) [inline recipes/elfio/all/test_package/CMakeLists.txt:11]: Should this check also be in the recipe?

File: recipes/elfio/all/conanfile.py
Comment (prince-chrismc) [inline recipes/elfio/all/conanfile.py:28]:
```suggestion
        self.copy(pattern=os.path.join("elfio", "*.hpp"), src=os.path.join(self.source_folder, self._source_subfolder), dst="include")
```

File: recipes/elfio/all/conandata.yml
Comment (prince-chrismc) [inline recipes/elfio/all/conandata.yml:6]:
```suggestion
    url: https://github.com/serge1/ELFIO/releases/download/Release_3.8/elfio-3.8.tar.gz
```

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (a4z) [reply recipes/elfio/all/test_package/CMakeLists.txt:11]: don't think so, it is not building anything. C++11 is meanwhile the default, and for those who run into build issues, they will get a very helpful message anyway. 
So I think it's better to keep the recipe simple

File: recipes/elfio/all/conanfile.py
Comment (a4z) [reply recipes/elfio/all/conanfile.py:28]: beside doing it different, what is the advantage. The pattern `folder/*.extension` is used all over the place in recipes in the cci

File: recipes/elfio/all/conandata.yml
Comment (a4z) [reply recipes/elfio/all/conandata.yml:6]: I don't see a diff, whitespace at the end?  
or the empty line (think one is required)

Comment (prince-chrismc) [reply recipes/elfio/all/conandata.yml:6]: Yes its the whitespace

File: recipes/elfio/all/conanfile.py
Comment (prince-chrismc) [reply recipes/elfio/all/conanfile.py:28]: folder separator on windows is `\`

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (prince-chrismc) [reply recipes/elfio/all/test_package/CMakeLists.txt:11]: How does the consumer know it's a requirement? What if I have `settings.compiler.cppstd=03` for my project? There's no indication it might be a problem 🤔

File: recipes/elfio/all/conandata.yml
Comment (a4z) [reply recipes/elfio/all/conandata.yml:6]: strange, my editor config should prevent that ....

Comment (a4z) [reply recipes/elfio/all/conandata.yml:6]: no, there are no change, when I try the Commit suggestion button, it says there are noch changes ...

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (a4z) [reply recipes/elfio/all/test_package/CMakeLists.txt:11]: you will get a very descriptive error message when you compile your program. and you will deserve it ;-) 

you can not have this check in the recipe, you can only check if the compiler in the profile is 'modern' enough to have C++11 support.
At some point in time you need to be able to assume a modern (10+ years old) compiler that has C++11 support is used, and imho this point is already here ;-)

File: recipes/elfio/all/conanfile.py
Comment (a4z) [reply recipes/elfio/all/conanfile.py:28]: ok, even if I think it makes functionally no difference, I guess it is maybe the more clean way to do it like that

File: recipes/elfio/all/conandata.yml
Comment (prince-chrismc) [reply recipes/elfio/all/conandata.yml:6]: It'll get fix when the next version gets add... I'd rather see it merged 👾

**Review by prince-chrismc — DISMISSED** [review]

**Comment by a4z** [comment]
please help! I can't see what went wrong with the build, the change should not have caused any failures, since nothing dramatical has changed ....

File: recipes/elfio/all/test_package/example.cpp
Comment (Hopobcn) [inline recipes/elfio/all/test_package/example.cpp:12]: Maybe if `__linux__` it could read itself `argv[0]` and print all contents using `elfio_dump.hpp`

```suggestion
    }
#ifdef __linux__
    std::string target = argv[0];

    if (!reader.load(target)) {
        std::cerr << "error: not an ELF file\n";
        return 1;
    }

    dump::header         ( std::cout, reader );
    dump::section_headers( std::cout, reader );
    dump::segment_headers( std::cout, reader );
    dump::symbol_tables  ( std::cout, reader );
    dump::notes          ( std::cout, reader );
    dump::dynamic_tags   ( std::cout, reader );
    dump::section_datas  ( std::cout, reader );
    dump::segment_datas  ( std::cout, reader );
#endif
```

File: recipes/elfio/all/test_package/example.cpp
Comment (Hopobcn) [inline recipes/elfio/all/test_package/example.cpp:3]:
```suggestion
#include <elfio/elfio.hpp>
#include <elfio/elfio_dump.hpp>
#include <cstdlib>
#include <iostream>
```

File: recipes/elfio/all/test_package/example.cpp
Comment (a4z) [reply recipes/elfio/all/test_package/example.cpp:12]: I don not see a very huge value in adding that complexity.
All that needs to be tested is that the header is found, and that is done. So test of the package works fine, we are not going to test elfio itself, this is done by elfio

**Review by Hopobcn — DISMISSED** [review]

**Comment by uilianries** [comment]
let me restart

**Review by uilianries — APPROVED** [review]

File: recipes/elfio/all/conandata.yml
Comment (SpaceIm) [inline recipes/elfio/all/conandata.yml:6]:
```suggestion
```

File: recipes/elfio/all/conanfile.py
Comment (SpaceIm) [inline recipes/elfio/all/conanfile.py:28]:
```suggestion
        self.copy("COPYING", src=self._source_subfolder, dst="licenses")
        self.copy(pattern=os.path.join("elfio", "*.hpp"), src=self._source_subfolder, dst="include")
```

File: recipes/elfio/all/test_package/example.cpp
Comment (SpaceIm) [inline recipes/elfio/all/test_package/example.cpp:1]:
```suggestion
```

File: recipes/elfio/all/conanfile.py
Comment (SpaceIm) [inline recipes/elfio/all/conanfile.py:12]:
```suggestion
    settings = "compiler"
```

File: recipes/elfio/all/conanfile.py
Comment (SpaceIm) [inline recipes/elfio/all/conanfile.py:18]:
```suggestion

    def configure(self):
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, 11)

```

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (SpaceIm) [inline recipes/elfio/all/test_package/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.1)
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/elfio/all/test_package/conanfile.py
Comment (SpaceIm) [inline recipes/elfio/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake", "cmake_find_package_multi"
```

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (SpaceIm) [inline recipes/elfio/all/test_package/CMakeLists.txt:8]:
```suggestion
conan_basic_setup(TARGETS)

find_package(elfio REQUIRED CONFIG)

add_executable(example example.cpp)
target_link_libraries(example elfio::elfio)
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]
show usage of official elfio CMake imported target

File: recipes/elfio/all/conanfile.py
Comment (a4z) [reply recipes/elfio/all/conanfile.py:32]: what is the difference ?

File: recipes/elfio/all/test_package/conanfile.py
Comment (a4z) [reply recipes/elfio/all/test_package/conanfile.py:7]: why adding the cmake_find_package_multi to the test, will this ever be used?

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (a4z) [reply recipes/elfio/all/test_package/CMakeLists.txt:8]: nice, I assume you tested that this works, yes?

Comment (a4z) [reply recipes/elfio/all/test_package/CMakeLists.txt:8]: but why this way, there are 3 ways to link, ${CONAN_LIBS}, CONAN_PKG::elfio and elfio::elfio
there might be even more
and we can not use all, so why that and not any other

File: recipes/elfio/all/test_package/example.cpp
Comment (a4z) [reply recipes/elfio/all/test_package/example.cpp:1]: I don not know what this does, remove an empty line, so the next editor config will add it back?

File: recipes/elfio/all/conanfile.py
Comment (SpaceIm) [reply recipes/elfio/all/conanfile.py:32]: `self.source_folder` is not required

File: recipes/elfio/all/test_package/conanfile.py
Comment (SpaceIm) [reply recipes/elfio/all/test_package/conanfile.py:7]: yes, by find_package

File: recipes/elfio/all/conanfile.py
Comment (a4z) [reply recipes/elfio/all/conanfile.py:32]: just shorter and relative to the current location? 
Ok

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (SpaceIm) [reply recipes/elfio/all/test_package/CMakeLists.txt:8]: all these ways are fine. but this one is the best for transparent integration.

Comment (a4z) [reply recipes/elfio/all/test_package/CMakeLists.txt:8]: actually, I think I like CONAN_PKG::elfio best, 
this is also what I mostly use, 
to be sure it's not found outside of conan, what could happen with elfio::elfio , right? 

so to test the package, I think ${CONAN_LIBS}, CONAN_PKG::elfio  are better suited

File: recipes/elfio/all/test_package/conanfile.py
Comment (a4z) [reply recipes/elfio/all/test_package/conanfile.py:7]: no, because we do not include the multi file in the cmake file, 
afaik this needs to query of a multi generator is used, and then include an other file
I do not think that adds any value to the test of the package

File: recipes/elfio/all/test_package/CMakeLists.txt
Comment (SpaceIm) [reply recipes/elfio/all/test_package/CMakeLists.txt:8]: > to be sure it's not found outside of conan, what could happen with elfio::elfio , right?

No

**Comment by a4z** [comment]
I really appreciate some of the feedback from this PR, but there is a moment where we do not talk about useful feedback anymore but just about doing things different, and it becomes stressfull and takes way too much time, of my time but also that from other reviewers, since each change requires a new review of 3 people.

I have now pulled in some of the latest changes, but for the last time.

If there is nothing that improves functionality or fixes errors, we either go with this or I close the PR and leave it to some one else to bring up again.

I am sorry for having to say this, but this PR starts to take much more or my energy than it should.

File: recipes/elfio/all/test_package/conanfile.py
Comment (SpaceIm) [reply recipes/elfio/all/test_package/conanfile.py:7]: config files are not "included", they are discovered by find_package. This modification is required if you accept the next modification in test_package.

**Comment by prince-chrismc** [comment]
I understand and hear your frustration, but if you could please take a step back and note we have 100s of PRs a month that we review, that does not include the 3 or 4 passes, and that number is growing. You have made some **incredible** contributions and all it it is appreciated. The challenge for us is the scalability of this project; the more variation the exponential growth in work.

@Croydon said it well recently https://github.com/kongaskristjan/fire-hpp/issues/9#issuecomment-722746034... We have our list of items to check and we mechanically go through them. Each of us is different but we are all looking to make this project a success.

🏆 Thank you for your contributions and working with us. Please keep opening PRs.

**Comment by a4z** [comment]
Chris, I think you miss understood me, sorry for being not clear enough.

I like to accept change requests, but at some moment, when change requests do not fix issues, or bring real improvements but just do things different, it becomes difficult.

I do not want to ignore the person who made the change request, because obviously there has been thought and effort invested,  on the other hand, accepting a request invalidates done reviews and that means means higher the load on you, because you need to again review the PR.
So actually, my indentions was/is to keep the workload lower and manageable, and therefore I asked to not make change requests that do just things different.

**Review by prince-chrismc — COMMENTED** [review]
There's two changes requested, which are to ensure we are correctly matching and generating the cmake files consumers expect. It's critical the cmake generators are as transparent as possible.

https://github.com/conan-io/conan-center-index/pull/3390/files?file-filters%5B%5D=.py&file-filters%5B%5D=.txt#r518705467
https://github.com/conan-io/conan-center-index/pull/3390/files?file-filters%5B%5D=.py&file-filters%5B%5D=.txt#r518704940

These are not about doing it _differently_ but doing it **right**, which I suspect you'll agree are worth it since there's no reviews that will be dismissed.

**Comment by a4z** [comment]
this PR is already so chaotic that I can not even see what you are linking to , @prince-chrismc 
For such a simple header only package an achievement.

If you mean the parts from the testing package, its in the comments why this is not a good idea to do so, this is the testing package and not your app consuming this package. So please  understand the difference what the purpose of the test package actually is.

In general
The package worked already fine at from the very the begin, it had enough review to be pulled in.

Then I pulled in a lot of things and invalidated the existing reviews at least for 3 times. Now I am done with that.

All the improvements should have been a different pr on top of that, not playing ping pong in this review. 

So, if you can not point to a bug in the package, than please, let it be and do not wast my time with further comment.

If you want additional improvements, contribute them on top of this. Wait until the package is in, and than contribute improvements, if you think they are important. Thanks.

**Comment by axalon900** [comment]
@a4z As someone who got very frustrated by this sort of thing recently with the fire-hpp PR, I sympathize. I think a lot of my frustration stems from the fact that the standards for a recipe are a moving target and the instructions are split between the FAQ, "how to add packages", the hooks, and plain issues. I think we could use a CCI-conforming template or a more detailed howto since stuff like the standards for writing a `test_package/` are basically missing. So you spend a lot of time getting it working and right, then get slapped down by what sounds like bikeshedding nonsense.

I don't mean this as an excuse for anything, but bear in mind the CCI project itself is still in "early access". While Conan itself is stable, the exact rules and regulations for recipe formatting in CCI in particular are in flux and sometimes underdefined. A lot of these seemingly arbitrary things come out organically from other pull requests which can give the impression that individual reviewers are out here pushing their pet preferences on you for whatever neurotic reason. I don't think anyone's being intentionally malicious, it's just that when you have to review lots and lots of mostly-similar things it makes it easier on everyone to have this or that be uniform. My suggestion in dealing with this sort of thing in general is to type out a rant, _don't submit it_, then just walk away and do something else. Once you come back to it you're usually not as heated and can turn that rant into some constructive feedback.

On the other hand, going through the trouble to make a request to remove blank lines is a _bit_ ridiculous. It's also happened in the past that reviewers request changes that are only tangentially related to the PR, they haven't tested it locally, and then they actually break the build, which is unacceptable. This isn't targeted to the CCI folks in particular, but I think reviewers really could be more cognizant of the fact that when you suggest things, the author will interpret that to mean "I will block this PR over this" unless you otherwise specify that it's just a suggestion for an improvement. And if that suggestion is, like, _whitespace_, wouldn't you get mad? It took me a long time to realize why I would get annoyed at some things but not others to the confusion of the reviewer who didn't mean it like that at all.

My two cents from my recent experiences as well as from experience at work, etc.

**Review by SSE4 — APPROVED** [review]

**Comment by SSE4** [comment]
I agree, in general, if PR passes CI (which implies passing hooks), and there is nothing really suspicious and blocking, it should be accepted.
trivial things like whitespace changes, formatting, double vs single quotes are really minor.
it's okay if the first PR doesn't support all platforms or compilers, all the conan generators or doesn't provide components, etc. all these functional improvements could be added in another iteration.
I also believe, that for the recuring things we should add automated checks to the hooks, if possible.
it guarantees the PR submitted will already have these things done correctly, and also reviewers will need to pay attention to fewer things, which will make the review process faster.

**Review by uilianries — APPROVED** [review]
LGTM

**Comment by SpaceIm** [comment]
There were minor fixes like CMake min version and check for min C++ standard, so I took the opportunity to also fix empty lines at the end of files. Accept this last suggestion takes what? 1s? I can't realize I have to answer to this kind of complains.

**Review by SpaceIm — APPROVED** [review]

**Comment by a4z** [comment]
> There were minor fixes like CMake min version and check for min C++ standard, .... 
And these where well appreciated improvements that I happily pulled in, with a thank you @SpaceIm

**Pull request merged**
